src/modules/fees/RadixTokenFeeCalculator.ts

```diff
@@ -0,0 +1,154 @@
+import { RadixAtom, RadixFixedSupplyTokenDefinitionParticle, RadixMutableSupplyTokenDefinitionParticle, RadixParticle, RadixSpin } from '../atommodel'
+import BN from 'bn.js'
+
+const bnZERO = new BN(0)
+
+const assertFeeDoesNotOverflow = (fee: BN) => {
+    const uint256Max = new BN(2).pow(new BN(256)).sub(new BN(1))
+
+    if (fee.gt(uint256Max)) {
+        throw new Error(`Fee cannot be larger than UInt256 max value`)
+    }
+}
+
+interface FeeEntry {
+    feeForAtom(atom: RadixAtom, feeSize: number, outputs: Set<RadixParticle>): BN
+}
+
+const perBytesFeeEntry = (units: number, threshold: number, fee: BN): FeeEntry => {
+
+    if (units === undefined || units === null) {
+        throw new Error(`units must not be null`)
+    }
+
+    if (threshold === undefined || threshold === null) {
+        throw new Error(`threshold must not be null`)
+    }
+
+    if (fee === undefined || fee === null) {
+        throw new Error(`Fee must not be null`)
+    }
+
+    if (units <= 0) {
+        throw new Error(`Units must be positive, but got: ${units}`)
+    }
+
+    if (threshold < 0) {
+        throw new Error(`Threshold must be non-negative, but got: ${threshold}`)
+    }
+
+    return {
+        feeForAtom: (atom: RadixAtom, feeSize: number, outputs: Set<RadixParticle>): BN => {
+            const numberOfUnits = feeSize / units
+            if (numberOfUnits <= threshold) {
+                return bnZERO
+            }
+            const overThresholdUnits = numberOfUnits - threshold
+            const feeTotal = fee.mul(new BN(overThresholdUnits))
+            assertFeeDoesNotOverflow(feeTotal)
+            return feeTotal
+        },
+    }
+}
+
+export type Class<T> = new(...args: any[]) => T
+
+export const filterOutParticlesOfTypeFromSet = (particles: Set<RadixParticle>, particleType: Class<RadixParticle>): RadixParticle[] => {
+    return filterOutParticlesOfTypeFromArray(Array.from(particles), particleType)
+}
+
+export const filterOutParticlesOfTypeFromArray = (particles: RadixParticle[], particleType: Class<RadixParticle>): RadixParticle[] => {
```

Comment:
The phrase "filterOut" suggests that particles of the given type are removed, whereas the opposite seems to be true here.

Reply:
mm singleOut is better